src/passes/OptimizeInstructions.cpp

```diff
@@ -1054,21 +1054,33 @@ struct OptimizeInstructions
   }
 
   void visitLocalSet(LocalSet* curr) {
-    //   (local.tee (ref.as_non_null ..))
-    // can be reordered to
-    //   (ref.as_non_null (local.tee ..))
-    // if the local is nullable (which it must be until some form of let is
-    // added). The reordering allows the ref.as to be potentially optimized
-    // further based on where the value flows to.
-    if (curr->isTee()) {
-      if (auto* as = curr->value->dynCast<RefAs>()) {
-        if (as->op == RefAsNonNull &&
-            getFunction()->getLocalType(curr->index).isNullable()) {
+    if (auto* as = curr->value->dynCast<RefAs>()) {
+      if (as->op == RefAsNonNull &&
+          getFunction()->getLocalType(curr->index).isNullable()) {
+        //   (local.tee (ref.as_non_null ..))
+        // can be reordered to
+        //   (ref.as_non_null (local.tee ..))
+        // if the local is nullable anyhow. The reordering allows the ref.as to
+        // be potentially optimized further based on where the value flows to.
+        if (curr->isTee()) {
           curr->value = as->value;
           curr->finalize();
           as->value = curr;
           as->finalize();
           replaceCurrent(as);
```

Comment:
Doesn't this transformation on tees inhibit the specialization of the local type just like we are trying to avoid in this PR?

Reply:
Good point, yeah, the pre-existing code does have that issue. I think it was written before we have LocalSubtyping...

Updated the PR to do this consistently.